mistralrs-core/src/utils/varbuilder_utils.rs

```diff
@@ -177,3 +69,67 @@ pub(crate) fn load_preload_adapters<'a>(
         Ok(None)
     }
 }
+
+// Presently this logic only needs to diverge for X-LoRA support via `into_name_key_pairs()`
+trait LoadTensors {
+    fn load_tensors_from_path(
+        path: &PathBuf,
+        device: &Device,
+        dtype: DType,
+        is_silent: bool,
+    ) -> Result<HashMap<String, Tensor>> {
+        let tensors = unsafe { candle_core::safetensors::MmapedSafetensors::new(path)? };
+
+        // Extracts the tensor name and processes it, filtering tensors and deriving the key name:
+        let names_only = tensors.tensors().into_iter().map(|(name, _)| name);
+        let iter = Self::into_name_key_pairs(names_only);
+
+        // Take the filtered list of tensors to load, store with derived lookup key:
+        let mut loaded_tensors = HashMap::new();
+        for (load_name, key_name) in iter.with_progress(is_silent) {
+            let tensor = tensors
+                .load(&load_name, device)?
+                // TODO: Seems redundant? Tensor was just loaded to this device?
+                // .to_device(&device)?
+                .to_dtype(dtype)?;
+
+            loaded_tensors.insert(key_name, tensor);
+        }
+
+        Ok(loaded_tensors)
+    }
+
+    fn into_name_key_pairs(
+        tensors: impl Iterator<Item = String>,
+    ) -> impl Iterator<Item = (String, String)> {
+        tensors.map(|name| {
+            let new_name = name.replace("base_model.model.model", "model");
+
+            (name, new_name)
+        })
+    }
+}
+
+struct Common {}
+impl LoadTensors for Common {}
+
+struct XLora {}
+impl LoadTensors for XLora {
+    fn into_name_key_pairs(
+        tensors: impl Iterator<Item = String>,
+    ) -> impl Iterator<Item = (String, String)> {
+        let expectation = "tensor name `{new_name}` should have substring `.lora`";
+
+        tensors
+            .filter(|name| !name.contains("internal_xlora_classifier"))
+            .enumerate()
+            .map(|(i, name)| {
+                let mut new_name = name.replace("base_model.model.model", "model");
+                // TODO: Add better context to describe intent / requirement:
+                let pos = new_name.find(".lora").expect(expectation);
+                new_name.insert_str(pos + 7, &format!(".{i}"));
+
+                (name, new_name)
+            })
+    }
+}
```

Comment:
~~This is a little different in that I `enumerate()` after excluded entries (_previously this was a `continue`, but that'd increment `i`_)~~ Whoops, `i` was only incremented per path, I've made a mistake here 😓

Reply:
> * Flip order of `.filter` and `.enumerate` (call enumerate before filter to keep old functionality)

That won't restore it, the `i` came before this logic, this is the inner loop of the existing approach. But the `i` was per `path`, not per entry from `tensors`.

I'll need to access that state from where I'm calling into the trait. So I'm going to add an `index` field to the `XLora` struct. Feel free to suggest a better name 😅

Reply:
Oh ok. What about `adapter_index`?